[ParseableInterface] Fix printing for properties with initializers and observers

harlanhaskins opened the pull request:
Instead of only printing through the pattern binding, and potentially missing stored properties with property observers, defer to the pattern binding for all stored properties, and print accessors if applicable while printing pattern bindings.

Fallout from rdar://43815861

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2047,6 +2053,27 @@ void PrintAST::visitExtensionDecl(ExtensionDecl *decl) {
+
+  if (Options.VarInitializers) {
+    // FIXME: Implement once we can pretty-print expressions.
+  }
```
Hm, this `if` should probably be guarding the next section…

harlanhaskins replied:
Yeah, this was left over from the original implementation. I'll just go ahead and change it. Should we even have the option to turn it off, now that var initializers are part of the public interface anyway?

jrose-apple replied:
I…guess not? Although I suppose we could have an option that chooses between "all of them" and "only the ones that are public".

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2406,21 +2415,35 @@ void PrintAST::visitVarDecl(VarDecl *decl) {
+
+  {
+    llvm::SaveAndRestore<bool>
+      printAsIUO(Options.PrintOptionalAsImplicitlyUnwrapped);
```
Can you add a comment stating why this whole mess is necessary?

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2406,21 +2415,35 @@ void PrintAST::visitVarDecl(VarDecl *decl) {
+    if (pbd && !Options.ExplodePatternBindingDecls) {
+      auto &entry = pbd->getPatternEntryForVarDecl(decl);
+      printPatternBindingEntry(entry);
+      printAccessors(decl);
```
This seems like it'll print the same entry twice when there are two VarDecls in the same entry.

harlanhaskins replied:
It might just be more profitable and cleaner to figure out when we print a single VarDecl without going through the pattern binding. That's what's really happening here.

jrose-apple commented on test/ParseableInterface/stored-properties.swift:
```diff
@@ -52,6 +52,15 @@ public struct HasStoredProperties {
   private var privateVar: Bool
 
+  // CHECK: @_hasStorage @_hasInitialValue public var storedWithObserversInitialValue: [[INT]] {
+  // RESILIENT: {{^}}  @_hasInitialValue public var storedWithObserversInitialValue: [[INT]] {
```
That `_hasInitialValue` is suspicious. We shouldn't present that knowledge in a resilient struct. (And arguably not in a fragile struct either, but let's not worry about that right now.)

harlanhaskins replied:
That makes sense -- that's the existing behavior, so I'll fix that in a follow-up patch.

harlanhaskins committed:
[ParseableInterface] Guard printing var initializers

Don’t print var initializers if `Options.VarInitializers` is false.

harlanhaskins committed:
[ParseableInterface] Clarify call to print accessors for single-entry pattern bindings

harlanhaskins committed:
[ParseableInterface] Go through pattern binding for all variables with storage

Instead of only printing through the pattern binding, and potentially missing stored properties with property observers, defer to the pattern binding for all stored properties, and print accessors if applicable while printing pattern bindings.

harlanhaskins commented:
@jrose-apple This should be much simpler.

jrose-apple commented:
Your test case got lost.

harlanhaskins commented:
@jrose-apple Thanks!

harlanhaskins commented:
@swift-ci please test and merge

harlanhaskins commented:
@swift-ci please test
